Require repo disambiguation for secret commands

williammartin opened the pull request:
## Description

Fixes https://github.com/cli/cli/issues/4688
Supersedes https://github.com/cli/cli/pull/9083

## Testing

There is a new acceptance test to run through the non-interactive cases:

```
➜  cli git:(wm/add-remote-check-to-secret) set -o pipefail && GH_ACCEPTANCE_SCRIPT=secret-require-remote-disambiguation.txtar GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -coverprofile=coverage.out -coverpkg=./... -run ^TestSecrets$ github.com/cli/cli/v2/acceptance | tparse --all go test
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │                       TEST                       │             PACKAGE               │
│─────────┼─────────┼──────────────────────────────────────────────────┼───────────────────────────────────│
│  PASS   │   11.45 │ TestSecrets/secret-require-remote-disambiguation │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestSecrets                                      │ github.com/cli/cli/v2/acceptance  │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │ 12.16s  │ github.com/cli/cli/v2/acceptance │ 10.2% │  2   │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

The other secret A/C tests that are expected to pass, pass.

There are some scripts in https://github.com/cli/cli/commit/0c9b6ed0ad02cc24b7252eccf791ccf7d8cb925b that demonstrate the behaviour. Most importantly, interactive prompting:

```sh
#!/usr/bin/env sh
_cli=/Users/williammartin/workspace/cli/
_tmp=/tmp
_org=williammartin-test-org
_repo=gh-some-repo
_gh="$_cli/bin/gh"
_pwd=$(pwd)
set -e
# Uncomment this to echo expanded commands:
#   set -x
cleanup () {
    ARG=$?
    cd $_pwd
    rm -rf "$_tmp/$_repo"
    $_gh repo delete --yes $_repo
    $_gh repo delete --yes "$_org/$_repo"
    exit $ARG
}
trap cleanup EXIT
cd $_tmp
_upstream=$($_gh repo create --private --add-readme $_repo)
$_gh repo fork --clone --org $_org $_upstream
cd $_repo

echo "setting secret..."
$_gh secret set ACCEPTANCE_KEY

echo "listing secrets..."
$_gh secret list

echo "deleting secret..."
$_gh secret delete ACCEPTANCE_KEY
echo 'PASS'
```

```
➜ ./multiple-git-remotes-interactive-prompting.sh
✓ Created fork williammartin-test-org/gh-some-repo
Cloning into 'gh-some-repo'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (3/3), done.
From https://github.com/williammartin/gh-some-repo
 * [new branch]      main       -> upstream/main
✓ Cloned fork
! Repository williammartin/gh-some-repo set as the default repository. To learn more about the default repository, run: gh repo set-default --help
setting secret...
! Multiple remotes detected. Due to the sensitive nature of secrets, requiring disambiguation.
? Select a base repo williammartin/gh-some-repo
? Paste your secret: ******

✓ Set Actions secret ACCEPTANCE_KEY for williammartin/gh-some-repo
listing secrets...
! Multiple remotes detected. Due to the sensitive nature of secrets, requiring disambiguation.
? Select a base repo williammartin/gh-some-repo
deleting secret...
! Multiple remotes detected. Due to the sensitive nature of secrets, requiring disambiguation.
? Select a base repo williammartin/gh-some-repo
✓ Deleted Actions secret ACCEPTANCE_KEY from williammartin/gh-some-repo
PASS
✓ Deleted repository williammartin/gh-some-repo
✓ Deleted repository williammartin-test-org/gh-some-repo
```

wingleung committed:
Update docs for set-default

NOTE: gh does not use the default repository for managing repository and environment secrets.

wingleung committed:
Update setdefault test

wingleung committed:
Add remote check to secret commands

Co-authored-by: William Martin <williammartin@github.com>

wingleung committed:
Add tests for secret commands

Co-authored-by: William Martin <williammartin@github.com>

wingleung committed:
Prompt for secret commands

Co-authored-by: William Martin <williammartin@github.com>

williammartin committed:
Always prompt on secret set when multiple remotes

williammartin committed:
Move secret repo validation into secrets subpackage

williammartin committed:
Use smarter base repo funcs for secret commands

williammartin committed:
Rename secret BaseRepo func

williammartin committed:
Remove validated TODO and add review warning

williammartin commented on pkg/cmd/secret/set/set.go:
```diff
@@ -187,6 +199,7 @@ func setRun(opts *SetOptions) error {
 		if err != nil {
 			return err
 		}
+
```
TODO: Probably move the base repo resolution above the secret prompts so that it happens before the user goes to any trouble:

This is not great:

```
➜ ~/workspace/cli/bin/gh secret set foo
? Paste your secret: ***

? Select a base repo  [Use arrows to move, type to filter]
  williammartin-test-org/test-repo
> williammartin/test-repo
```

jtmcg commented on pkg/cmd/secret/delete/delete.go:
```diff
@@ -46,8 +46,19 @@ func NewCmdDelete(f *cmdutil.Factory, runF func(*DeleteOptions) error) *cobra.Co
 			opts.BaseRepo = f.BaseRepo
+			if !cmd.Flags().Changed("repo") {
+				// If they haven't specified a repo directly, then we will wrap the BaseRepoFunc in one that error if
+				// there might be multiple valid remotes.
```
Nit: typo

```suggestion
				// If they haven't specified a repo directly, then we will wrap the BaseRepoFunc in one that errors if
				// there might be multiple valid remotes.
```

jtmcg commented on pkg/cmd/secret/delete/delete.go:
```diff
@@ -46,8 +46,19 @@ func NewCmdDelete(f *cmdutil.Factory, runF func(*DeleteOptions) error) *cobra.Co
+			if !cmd.Flags().Changed("repo") {
+				// If they haven't specified a repo directly, then we will wrap the BaseRepoFunc in one that error if
+				// there might be multiple valid remotes.
+				opts.BaseRepo = shared.RequireNoAmbiguityBaseRepoFunc(opts.BaseRepo, f.Remotes)
+				// But if we are able to prompt, then we will wrap that up in a BaseRepoFunc that can prompt the user to
+				// resolve the ambiguity.
+				if opts.IO.CanPrompt() {
+					opts.BaseRepo = shared.PromptWhenAmbiguousBaseRepoFunc(opts.BaseRepo, f.Prompter)
+				}
+			}
```
**Question**

What is the reasoning behind putting this here instead of adding it to the [factory](https://github.com/cli/cli/blob/21a1f8e8a9201143b11cb8b8e5d78bc163839c60/pkg/cmd/factory/default.go#L72-L80)?

Disclaimer: I haven't investigated the impact of that throughout the rest of the app, but disambiguating baseRepo everywhere sounds like a good idea in theory...

andyfeller replied:
If I would predict Will's response, there is value in having command-specific logic captured within the relevant command.

Applying these changes within the factory to apply everywhere would be quite a bit of scope creep and it would also potentially make testing somewhat complicated whereas keeping it here allows tests to override `opts.BaseRepo` and still have that stub/mock wrapped the same. 🤷

williammartin replied:
> What is the reasoning behind putting this here instead of adding it to the [factory](https://github.com/cli/cli/blob/21a1f8e8a9201143b11cb8b8e5d78bc163839c60/pkg/cmd/factory/default.go#L72-L80)?

It's a breaking change, tightly scoped to secret commands (we backed out variable commands), and I want there to be some friction in order to use it elsewhere.

> Disclaimer: I haven't investigated the impact of that throughout the rest of the app, but disambiguating baseRepo everywhere sounds like a good idea in theory...

I didn't even read this before responding to the previous statement, and it's a perfect example 😅.

We could document it of course but I also don't think there's clear value pulling it up that makes that tradeoff worth it.

---

I could be convinced in some shared logic for the secret package. Unfortunately, I think it would have to live in some kind of `PreRun`, in the `secret` command and I think that would be pretty hard to find (and not a pattern we have anywhere else).

I'm open to ideas though.

@jtmcg I don't understand this comment:

> That way all the functionality is hoisted higher, you remove the need for repetitive testing, and you can rely on mocking in the command tests for the base repo instead of writing the tests you have here to validate the behavior.

Something needs to test that the correct BaseRepo func has been chosen. Where do you think that would go?

jtmcg replied:
For sure, your explanation makes sense 👍 

> Something needs to test that the correct BaseRepo func has been chosen. Where do you think that would go?

This would be, theoretically, scoped to the factory if it got moved up. Then the downstream consumers don't care which BaseRepo function got chosen as long as they're getting one that works. It's really moot, though, given the reasons you've provided to not move it to the factory

jtmcg replied:
> I could be convinced in some shared logic for the secret package. Unfortunately, I think it would have to live in some kind of PreRun, in the secret command and I think that would be pretty hard to find (and not a pattern we have anywhere else).

I think that this might remove some of the lack of understanding I had in the tests (called out in the below [comment](https://github.com/cli/cli/pull/10209/files#r1910405405)). Then you're testing that the correct function is selected in that shared spot instead of repeating it 3 times in `delete`, `set`, and `list`.

That said, I don't have any ideas off hand on how to do this without trying it. I'll take a look if I have time, but I'm happy to accept this implementation given my comment above

williammartin replied:
Think this is resolved in light of confusion being cleared up and https://github.com/cli/cli/pull/10209#discussion_r1916667343

jtmcg commented on pkg/cmd/secret/delete/delete_test.go:
```diff
@@ -120,6 +125,108 @@ func TestNewCmdDelete(t *testing.T) {
+		},
+	}
+
+	tests := []struct {
```
**Food for Thought**

We may be missing a test case:

"when there is no repo flag provided, no prompting, but the base repo has no ambiguity"

If I understand correctly, this can happen when there is a single remote, resulting in no `AmbiguityBaseRepoError`

andyfeller replied:
Would that be a totally separate test given this test seems to hinge on the same 2 remotes for each test case?

williammartin replied:
Hmmm, ok I read this comment last after the other ones and I think maybe there's a misunderstanding.

These tests, and the same ones in `list` and `set` are validating that **the correct base repo func is selected**.

They are not testing the specifics of how those base repo funcs work.

> "when there is no repo flag provided, no prompting, but the base repo has no ambiguity"

The last part of this has no bearing on the selection of BaseRepo func.

---

Maybe it would be clearer if I renamed these tests:

`TestNewCmdDeleteBaseRepoFuncSelection`?

williammartin replied:
Separately, I'm also not wild about these tests, since (and evidenced by your confusion) they have overlap with the actual tests. The problem is that there is an artificial seam between `PreRun` and `Run` and it's not really viable to test function equality that the "correct function was chosen".

I did look into using some generics to make a shared test case that would be usable for both commands but the readability was severely diminished.

jtmcg replied:
Ah, yes, I think you're right about my misunderstanding... When I'm faced with this kind of test overlap it usually signals to me that pulling out another function to test might be worthwhile. Like, the command doesn't care how it gets BaseRepo, as long as it can, so adding something like

```golang
    opts.BaseRepo = shared.DetermineBaseRepoFunction(opts.BaseRepo, f.Prompter, f.Remotes, opts.IO.CanPrompt())
```

and testing `DetermineBaseRepoFunction` instead of these three seems practical. As stated above, though, I don't know how practical that is, so happy to accept this as-is

williammartin replied:
I suppose we could do that, but then we'd need a way to inject a spy to ensure the correct function is called, and I don't really want that on the public interface.

If it's ok with you I'd like to leave it alone for now, this has been quite a long process to get through already.

jtmcg commented on pkg/cmd/secret/list/list_test.go:
```diff
@@ -101,6 +105,108 @@ func Test_NewCmdList(t *testing.T) {
+		},
+	}
+
+	tests := []struct {
```
**Food for Thought**

Same as above - we may be missing the test case when there's a single remote.

jtmcg commented on pkg/cmd/secret/set/set_test.go:
```diff
@@ -221,6 +224,108 @@ func TestNewCmdSet(t *testing.T) {
+		},
+	}
+
+	tests := []struct {
```
**Food for thought**

Same as above - we may be missing the test case for a single remote

jtmcg commented on pkg/cmd/secret/shared/base_repo_test.go:
```diff
@@ -0,0 +1,239 @@
+func TestRequireNoAmbiguityBaseRepoFunc(t *testing.T) {
+	t.Parallel()
+
+	t.Run("succeeds when there is only one remote", func(t *testing.T) {
```
**Thinking out loud**

This _may_ replace the need for the missing test cases I mentioned above, but part of me is getting a 👃 about the similarity of all the tests for `delete`, `list`, and `set` being essentially the same...

I think this alludes to the factory as maybe being the better spot for where this disambiguation behavior lives.

williammartin replied:
I think this is broadly addressed by my comment above in terms of getting on the same page about what you're proposing. https://github.com/cli/cli/pull/10209#discussion_r1910400447

jtmcg reviewed (commented):
Clever implementation, but I'm wondering if hoisting this even higher and into the [factory's `BaseRepo` function](https://github.com/cli/cli/blob/21a1f8e8a9201143b11cb8b8e5d78bc163839c60/pkg/cmd/factory/default.go#L72-L80) might be the right choice. I'm not sure about the cost of that throughout the rest of the app, but the amount of repetition in the tests are giving me pause.

Maybe it isn't expansion of `BaseRepo` but instead the introduction of an `UnambiguousBaseRepo` function on the factory that secret delete, list, and set all use in place of `BaseRepo`. That way all the functionality is hoisted higher, you remove the need for repetitive testing, and you can rely on mocking in the command tests for the base repo instead of writing the tests you have here to validate the behavior.

williammartin committed:
Add acceptance test for secrets remote disambiguation

williammartin committed:
Move secret base repo prompting earlier

williammartin committed:
Secret base repo prompting should not use resolved remote

This is because the secret commands don't use the SmartBaseRepo behaviour, and therefore
don't care about the resolved remote.

williammartin committed:
Print informative message before prompting for secret repo

williammartin committed:
Change wording on secret repo prompt

williammartin committed:
Fix typo in secret base repo selection comment

andyfeller commented on pkg/cmd/secret/delete/delete.go:
```diff
@@ -88,6 +99,14 @@ func removeRun(opts *DeleteOptions) error {
 		return err
 	}
 
+	var baseRepo ghrepo.Interface
+	if secretEntity == shared.Repository || secretEntity == shared.Environment {
+		baseRepo, err = opts.BaseRepo()
+		if err != nil {
+			return err
+		}
+	}
+
```
I assume this was moved higher up so any ambiguity errors are raised earlier before the other ways this can error.

williammartin replied:
Yeh, similar to https://github.com/cli/cli/pull/10209#discussion_r1909138595 I just wanted the repo disambiguation frontloaded to avoid going through a bunch of steps only to discover this at the end.

andyfeller commented on pkg/cmd/secret/shared/base_repo_test.go:
```diff
@@ -0,0 +1,249 @@
+)
+
+func TestRequireNoAmbiguityBaseRepoFunc(t *testing.T) {
+	t.Parallel()
+
+	t.Run("succeeds when there is only one remote", func(t *testing.T) {
+		t.Parallel()
```
I think this would make a great topic of conversation about testing craftsmanship.

andyfeller reviewed (commented):
I'm giving this a preliminary LGTM having read the code, I'm going to build this locally and report back results before final approval.

andyfeller reviewed (approved):
Having manually put it through its paces, I think this is good to go.

### Testing notes

Building PR locally

```shell
$ cd Documents/workspace/cli/cli
$ gh pr checkout 10209          
Already on 'wm/add-remote-check-to-secret'
Your branch is up to date with 'origin/wm/add-remote-check-to-secret'.
Already up to date.

$ make
go build -trimpath -ldflags "-X github.com/cli/cli/v2/internal/build.Date=2025-01-15 -X github.com/cli/cli/v2/internal/build.Version=v2.65.0-138-ge7ffb1e4 " -o bin/gh ./cmd/gh
``

Setting up upstream and fork repos for testing

```shell
$ ~/Documents/workspace/cli/cli/bin/gh repo create gh-acceptance-testing/10209-upstream --add-readme --private --clone 
✓ Created repository gh-acceptance-testing/10209-upstream on GitHub
  https://github.com/gh-acceptance-testing/10209-upstream
Cloning into '10209-upstream'...

$ ~/Documents/workspace/cli/cli/bin/gh repo fork gh-acceptance-testing/10209-upstream --org gh-acceptance-testing --fork-name 10209-fork
✓ Created fork gh-acceptance-testing/10209-fork
? Would you like to clone the fork? Yes
Cloning into '10209-fork'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (3/3), done.
From https://github.com/gh-acceptance-testing/10209-upstream
 * [new branch]      main       -> upstream/main
✓ Cloned fork
! Repository gh-acceptance-testing/10209-upstream set as the default repository. To learn more about the default repository, run: gh repo set-default --help
```

Confirming ambiguous prompt behavior

```shell
$ cd 10209-fork 

$ ~/Documents/workspace/cli/cli/bin/gh secret set SECRET --body "should be fork" 
! Multiple remotes detected. Due to the sensitive nature of secrets, requiring disambiguation.
? Select a repo  [Use arrows to move, type to filter]
> gh-acceptance-testing/10209-upstream
  gh-acceptance-testing/10209-fork
```

Creating secrets on upstream and fork for testing in workflow run

```shell
$ ~/Documents/workspace/cli/cli/bin/gh secret set SECRET --body "should be fork" --repo gh-acceptance-testing/10209-fork
✓ Set Actions secret SECRET for gh-acceptance-testing/10209-fork

$ ~/Documents/workspace/cli/cli/bin/gh secret set SECRET --body "should be upstream" --repo gh-acceptance-testing/10209-upstream
✓ Set Actions secret SECRET for gh-acceptance-testing/10209-upstream
```

Creating workflow to test secrets

```shell
$ cd ..                          
$ cd 10209-upstream 
$ mkdir -p .github/workflows
$ cat << EOF > .github/workflows/verify.yml
# This workflow is intended to assert the value of the GitHub Actions secret was set appropriately
name: Test Workflow Name
on:
  # Allow workflow to be dispatched by gh workflow run
  workflow_dispatch:

jobs:
  # This workflow contains a single job called "assert" that should only pass if the GitHub Actions secret value matches
  assert:
    runs-on: ubuntu-latest
    steps:
      - name: Assert secret value matches
        env:
          SECRET: ${{ secrets.SECRET }}
        run: |
          if [[ "$SECRET" == "should be upstream" ]]; then
            echo "GitHub Actions secret value states it should be the upstream value"
          elif [[ "$SECRET" == "should be fork" ]]; then
            echo "GitHub Actions secret value states it should be the fork value"
          else
            echo "GitHub Actions secret value does not match anything we expect; this is bad"
            exit 1
          fi
EOF

$ git add verify.yml
$ git commit -m "Test workflow"
$ git push
Enumerating objects: 6, done.
Counting objects: 100% (6/6), done.
Delta compression using up to 16 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (5/5), 762 bytes | 762.00 KiB/s, done.
Total 5 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
To https://github.com/gh-acceptance-testing/10209-upstream.git
   006dbc5..4c1d420  main -> main
```

Testing secret values with workflow in upstream

```shell
$ ~/Documents/workspace/cli/cli/bin/gh run view 12791122577 --repo gh-acceptance-testing/10209-upstream --log
assert	Set up job	2025-01-15T15:04:17.5302179Z Current runner version: '2.321.0'
assert	Set up job	2025-01-15T15:04:17.5340288Z ##[group]Operating System
assert	Set up job	2025-01-15T15:04:17.5341721Z Ubuntu
assert	Set up job	2025-01-15T15:04:17.5342556Z 24.04.1
assert	Set up job	2025-01-15T15:04:17.5343331Z LTS
assert	Set up job	2025-01-15T15:04:17.5344352Z ##[endgroup]
assert	Set up job	2025-01-15T15:04:17.5345387Z ##[group]Runner Image
assert	Set up job	2025-01-15T15:04:17.5346389Z Image: ubuntu-24.04
assert	Set up job	2025-01-15T15:04:17.5347395Z Version: 20250105.1.0
assert	Set up job	2025-01-15T15:04:17.5349449Z Included Software: https://github.com/actions/runner-images/blob/ubuntu24/20250105.1/images/ubuntu/Ubuntu2404-Readme.md
assert	Set up job	2025-01-15T15:04:17.5351932Z Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu24%2F20250105.1
assert	Set up job	2025-01-15T15:04:17.5353680Z ##[endgroup]
assert	Set up job	2025-01-15T15:04:17.5354509Z ##[group]Runner Image Provisioner
assert	Set up job	2025-01-15T15:04:17.5355565Z 2.0.414.1
assert	Set up job	2025-01-15T15:04:17.5356502Z ##[endgroup]
assert	Set up job	2025-01-15T15:04:17.5358706Z ##[group]GITHUB_TOKEN Permissions
assert	Set up job	2025-01-15T15:04:17.5361479Z Contents: read
assert	Set up job	2025-01-15T15:04:17.5362448Z Metadata: read
assert	Set up job	2025-01-15T15:04:17.5363318Z Packages: read
assert	Set up job	2025-01-15T15:04:17.5364702Z ##[endgroup]
assert	Set up job	2025-01-15T15:04:17.5369809Z Secret source: Actions
assert	Set up job	2025-01-15T15:04:17.5370966Z Prepare workflow directory
assert	Set up job	2025-01-15T15:04:17.5856623Z Prepare all required actions
assert	Set up job	2025-01-15T15:04:17.5997410Z Complete job name: assert
assert	Assert secret value matches	2025-01-15T15:04:17.7028911Z ##[group]Run if [[ "$SECRET" == "***" ]]; then
assert	Assert secret value matches	2025-01-15T15:04:17.7029780Z if [[ "$SECRET" == "***" ]]; then
assert	Assert secret value matches	2025-01-15T15:04:17.7030538Z   echo "GitHub Actions secret value states it should be the upstream value"
assert	Assert secret value matches	2025-01-15T15:04:17.7031334Z elif [[ "$SECRET" == "should be fork" ]]; then
assert	Assert secret value matches	2025-01-15T15:04:17.7032092Z   echo "GitHub Actions secret value states it should be the fork value"
assert	Assert secret value matches	2025-01-15T15:04:17.7032792Z else
assert	Assert secret value matches	2025-01-15T15:04:17.7033440Z   echo "GitHub Actions secret value does not match anything we expect; this is bad"
assert	Assert secret value matches	2025-01-15T15:04:17.7034181Z   exit 1
assert	Assert secret value matches	2025-01-15T15:04:17.7034574Z fi
assert	Assert secret value matches	2025-01-15T15:04:17.7258622Z shell: /usr/bin/bash -e {0}
assert	Assert secret value matches	2025-01-15T15:04:17.7259548Z env:
assert	Assert secret value matches	2025-01-15T15:04:17.7260120Z   SECRET: ***
assert	Assert secret value matches	2025-01-15T15:04:17.7260542Z ##[endgroup]
assert	Assert secret value matches	2025-01-15T15:04:17.7506265Z GitHub Actions secret value states it should be the upstream value
assert	Complete job	2025-01-15T15:04:17.7609521Z Cleaning up orphan processes
```

Syncing local fork to pickup workflow change

```shell
$ ~/Documents/workspace/cli/cli/bin/gh repo sync       
✓ Synced the "main" branch from "gh-acceptance-testing/10209-upstream" to local repository

$ git push
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
To https://github.com/gh-acceptance-testing/10209-fork.git
   006dbc5..6db6d5e  main -> main
```

Testing secret values with workflow in fork

```shell
$ ~/Documents/workspace/cli/cli/bin/gh workflow run verify.yml --repo gh-acceptance-testing/10209-fork
✓ Created workflow_dispatch event for verify.yml at main

To see runs for this workflow, try: gh run list --workflow=verify.yml

$ ~/Documents/workspace/cli/cli/bin/gh run list --repo gh-acceptance-testing/10209-fork               
STATUS  TITLE               WORKFLOW            BRANCH  EVENT              ID           ELAPSED  AGE                   
*       Test Workflow Name  Test Workflow Name  main    workflow_dispatch  12791264263  10s      less than a minute ago

$ ~/Documents/workspace/cli/cli/bin/gh run view 12791264263 --log --repo gh-acceptance-testing/10209-fork
assert	Set up job	2025-01-15T15:11:08.7937555Z Current runner version: '2.321.0'
assert	Set up job	2025-01-15T15:11:08.7969975Z ##[group]Operating System
assert	Set up job	2025-01-15T15:11:08.7970877Z Ubuntu
assert	Set up job	2025-01-15T15:11:08.7971579Z 24.04.1
assert	Set up job	2025-01-15T15:11:08.7972354Z LTS
assert	Set up job	2025-01-15T15:11:08.7972915Z ##[endgroup]
assert	Set up job	2025-01-15T15:11:08.7973627Z ##[group]Runner Image
assert	Set up job	2025-01-15T15:11:08.7974320Z Image: ubuntu-24.04
assert	Set up job	2025-01-15T15:11:08.7974948Z Version: 20250105.1.0
assert	Set up job	2025-01-15T15:11:08.7976335Z Included Software: https://github.com/actions/runner-images/blob/ubuntu24/20250105.1/images/ubuntu/Ubuntu2404-Readme.md
assert	Set up job	2025-01-15T15:11:08.7978054Z Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu24%2F20250105.1
assert	Set up job	2025-01-15T15:11:08.7979165Z ##[endgroup]
assert	Set up job	2025-01-15T15:11:08.7979889Z ##[group]Runner Image Provisioner
assert	Set up job	2025-01-15T15:11:08.7980606Z 2.0.414.1
assert	Set up job	2025-01-15T15:11:08.7981115Z ##[endgroup]
assert	Set up job	2025-01-15T15:11:08.7982801Z ##[group]GITHUB_TOKEN Permissions
assert	Set up job	2025-01-15T15:11:08.7985103Z Contents: read
assert	Set up job	2025-01-15T15:11:08.7985850Z Metadata: read
assert	Set up job	2025-01-15T15:11:08.7986468Z Packages: read
assert	Set up job	2025-01-15T15:11:08.7987448Z ##[endgroup]
assert	Set up job	2025-01-15T15:11:08.7990772Z Secret source: Actions
assert	Set up job	2025-01-15T15:11:08.7991927Z Prepare workflow directory
assert	Set up job	2025-01-15T15:11:08.8320617Z Prepare all required actions
assert	Set up job	2025-01-15T15:11:08.8415448Z Complete job name: assert
assert	Assert secret value matches	2025-01-15T15:11:08.9501281Z ##[group]Run if [[ "$SECRET" == "should be upstream" ]]; then
assert	Assert secret value matches	2025-01-15T15:11:08.9503026Z if [[ "$SECRET" == "should be upstream" ]]; then
assert	Assert secret value matches	2025-01-15T15:11:08.9504563Z   echo "GitHub Actions secret value states it should be the upstream value"
assert	Assert secret value matches	2025-01-15T15:11:08.9506369Z elif [[ "$SECRET" == "***" ]]; then
assert	Assert secret value matches	2025-01-15T15:11:08.9507927Z   echo "GitHub Actions secret value states it should be the fork value"
assert	Assert secret value matches	2025-01-15T15:11:08.9509304Z else
assert	Assert secret value matches	2025-01-15T15:11:08.9510576Z   echo "GitHub Actions secret value does not match anything we expect; this is bad"
assert	Assert secret value matches	2025-01-15T15:11:08.9512315Z   exit 1
assert	Assert secret value matches	2025-01-15T15:11:08.9513214Z fi
assert	Assert secret value matches	2025-01-15T15:11:08.9866996Z shell: /usr/bin/bash -e {0}
assert	Assert secret value matches	2025-01-15T15:11:08.9868123Z env:
assert	Assert secret value matches	2025-01-15T15:11:08.9868850Z   SECRET: ***
assert	Assert secret value matches	2025-01-15T15:11:08.9869374Z ##[endgroup]
assert	Assert secret value matches	2025-01-15T15:11:09.0102099Z GitHub Actions secret value states it should be the fork value
assert	Complete job	2025-01-15T15:11:09.0242915Z Cleaning up orphan processes
```
